pkg/master/thirdparty/thirdparty.go

```diff
@@ -166,25 +180,118 @@ func (m *ThirdPartyResourceServer) RemoveThirdPartyResource(path string) error {
 	return nil
 }
 
-func (m *ThirdPartyResourceServer) removeAllThirdPartyResources(registry *thirdpartyresourcedatastore.REST) error {
+func (m *ThirdPartyResourceServer) removeThirdPartyResourceData(group *metav1.APIGroup, resource string, registry *thirdpartyresourcedatastore.REST) error {
+	// Freeze TPR data to prevent new writes via this apiserver process.
+	// Other apiservers can still write. This is best-effort because there
+	// are worse problems with TPR data than the possibility of going back
+	// in time when migrating to CRD [citation needed].
+	registry.Freeze()
+
 	ctx := genericapirequest.NewDefaultContext()
 	existingData, err := registry.List(ctx, nil)
 	if err != nil {
 		return err
 	}
 	list, ok := existingData.(*extensions.ThirdPartyResourceDataList)
 	if !ok {
-		return fmt.Errorf("expected a *ThirdPartyResourceDataList, got %#v", list)
+		return fmt.Errorf("expected a *ThirdPartyResourceDataList, got %T", existingData)
 	}
-	for ix := range list.Items {
-		item := &list.Items[ix]
-		if _, _, err := registry.Delete(ctx, item.Name, nil); err != nil {
+
+	gvk := schema.GroupVersionKind{Group: group.Name, Version: group.PreferredVersion.Version, Kind: registry.Kind()}
+	if err := m.migrateThirdPartyResourceData(gvk, resource, list); err != nil {
+		// Migration is best-effort. Log and continue.
+		utilruntime.HandleError(fmt.Errorf("failed to migrate TPR data: %v", err))
+	}
+
+	for i := range list.Items {
+		item := &list.Items[i]
+
+		// Use registry.Store.Delete() to bypass the frozen registry.Delete().
+		if _, _, err := registry.Store.Delete(ctx, item.Name, nil); err != nil {
 			return err
 		}
 	}
 	return nil
 }
 
+func (m *ThirdPartyResourceServer) migrateThirdPartyResourceData(gvk schema.GroupVersionKind, resource string, dataList *extensions.ThirdPartyResourceDataList) error {
+	// Look for a matching CRD.
+	// CustomResourceDefinitionList does not implement the protobuf marshalling interface.
+	config := *m.genericAPIServer.LoopbackClientConfig
+	config.ContentType = "application/json"
+	crdClient, err := apiextensionsclient.NewForConfig(&config)
+	if err != nil {
+		return fmt.Errorf("can't create apiextensions client: %v", err)
+	}
+	crdList, err := crdClient.CustomResourceDefinitions().List(metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("can't list CustomResourceDefinitions: %v", err)
+	}
+	var crd *apiextensions.CustomResourceDefinition
+	for i := range crdList.Items {
+		item := &crdList.Items[i]
+		if item.Spec.Scope == apiextensions.NamespaceScoped &&
+			item.Spec.Group == gvk.Group && item.Spec.Version == gvk.Version &&
+			item.Status.AcceptedNames.Kind == gvk.Kind && item.Status.AcceptedNames.Plural == resource {
+			crd = item
+			break
+		}
+	}
+	if crd == nil {
+		// No matching CRD, so no migration necessary.
+		return nil
+	}
+
+	// Talk directly to CustomResource storage.
+	// We have to bypass the API server because TPR is shadowing CRD at this point.
```

Comment:
will a client observe the deletions? Thinking about the etcd operator deleting its storage volumes.